packages/typescript/internal/ts_project.bzl

```diff
@@ -98,6 +122,19 @@ def _calculate_root_dir(ctx):
     )
 
 def _ts_project_impl(ctx):
+    srcs = [src.path for src in ctx.files.srcs]
+    typings_out_dir = ctx.attr.declaration_dir or ctx.attr.out_dir
+
+    # Recalculate outputs inside the rule implementation.
+    # The outs are first calculated in the macro in order to try to predetermine outputs so they can be declared as
+    # outputs on the rule. This provides the benefit of being able to reference an output file with a label.
+    # However, it is not possible to evaluate files in outputs other rules such as filegroup, therefore the outs a
+    # recalculated here.
+    js_outs = _declare_outputs(ctx, _calculate_js_outs(srcs, ctx.attr.out_dir, ctx.attr.root_dir, ctx.attr.allow_js, ctx.attr.preserve_jsx, ctx.attr.emit_declaration_only))
```

Comment:
I'm trying to convince myself that the outputs we declare here will always include all the outputs which were predeclared. obviously they share the _calculate implementations. And I think Bazel will error if a predeclared output isn't created by any actions in the implementation. This seems simple :)

Reply:
The "predeclared_outpus" will only find src that end with ".ts" or ".tsx", if an input is a label pointing to a target, then they won't be predeclared. 
The predeclared outputs are provided in `ctx.outputs`, but Bazel is equally happen if you use `ctx.actions.declare_file` for the same file. It figures out it's a predeclared output file and returns the same file object.

Reply:
yeah that matches my understanding also, SGTM